### PR TITLE
feat: OCR features 

### PR DESCRIPTION
**Pull request by AnisZakari**
Ocr functions

### CODE REVIEW

File: make_ocr_features/make_OCR_df_and_preprocess.ipynb [hunk 1]
```diff
@@ -0,0 +1,1131 @@
...
+      "250000\n",
+      "500000\n",
+      "750000\n",
+      "1000000\n"
+     ]
+    }
+   ],
+   "source": [
+    "\"\"\"takes approx 7m30\"\"\"\n",
+    "txt_annotations_path = \"INPUT_datasets/txtannotations.jsonl.gzip\"\n",
```
Comment (streino) [inline make_ocr_features/make_OCR_df_and_preprocess.ipynb:669]: Where does that file come from? -> add to README
(I couldn't run notebook further, will do when I have the file)

**Review by streino — CHANGES_REQUESTED** [review]

**Comment by streino** [comment]
Forgot to mention in my review: how is this going to be added to model training?

My suggestion would be to integrate it in the [tdfs dataset generation](https://github.com/openfoodfacts/off-category-classification/blob/master/datasets/off_categories/off_categories.py) and make it available as a (or several) new feature(s).

In any case, the processed OCR data should be output in some way that's easy to process further (and can be released along with other data files?). So pickle probably isn't the best format.

It'd probably also be a good idea to keep line-level correspondence with original data files as well, so we don't have to do costly merge operations on codes if we want to join various feature files. This would pretty much be mandatory if we go the tdfs dataset route.

File: make_ocr_features/make_OCR_df_and_preprocess.ipynb [hunk 2]
```diff
@@ -0,0 +1,1070 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 28,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from collections import defaultdict\n",
+    "big_words_dict = defaultdict(str)\n",
```
Comment (streino) [inline make_ocr_features/make_OCR_df_and_preprocess.ipynb:610]: I'd move those vars in the next cell so they get properly reinitialized if next cell get re-run.

File: make_ocr_features/utils/OCR_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+### cleaner ###
+
+def text_cleaner(text:str) -> str:
+    """ takes a text as input and cleans it"""
+
+    dont_take = ["kj", "kcal", "kj", "total", "free", "net", "ingredients", "ingredient", "et", "de", "fat", "mg", "cg", "g", "kg", "ml", "cl", "l", "kl", "per", "pour", "valeur", "or", "le", "la", "dont", "consommer", "poids", "net", "www", "com", "which", "of", "wt"]
+    text_cleaned = text.replace("\n", " ") #remove line breaks
+    text_cleaned = re.sub(r'''(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’]))''', " ", text_cleaned) #remove websites
```
Comment (streino) [inline make_ocr_features/utils/OCR_preprocessing.py:64]: That's quite a regex! Is this something you borrowed (link to source?) or you built? I'll trust you on that one, it looks scary...

File: make_ocr_features/utils/OCR_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    text_cleaned = " ".join([w for w in text_cleaned.split() if (w.isalpha() and w.lower() not in dont_take)])
+    return text_cleaned
+
+
+###remove duplicates
+def remove_duplicates(text, get_keys = False):
+    """takes a text and removes the duplicated words.
+    It keeps the string's original casing if get_keys = True
+    """
+    D = {word.lower(): word  for word in str(text).split()}
```
Comment (streino) [inline make_ocr_features/utils/OCR_preprocessing.py:80]: Pep8: Avoid uppercase vars

File: make_ocr_features/utils/OCR_word_selection.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    else:
+        words_picked = " ".join([str(w) for w in df_text.split() if (str(w).lower() in word_list and freq_dict[str(w).lower()]>2)])
+    return remove_duplicates(words_picked)
+
+def get_freq_dict(text_series):
+    """ computes word frequencies in text_series"""
+    freq_dict = defaultdict(int)
+    for text in text_series:
+        for word in text.split():
+            w_lower = str(word).lower()
```
Comment (streino) [inline make_ocr_features/utils/OCR_word_selection.py:83]: Nitpick: I mentioned it in another comment as well: Why the `str()` call here (and other places)? Aren't we guaranteed that `text.split()` returns strings?

File: make_ocr_features/utils/OCR_word_selection.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+def get_big_words_from_image(barcode, df, keys = None, word_count_limit = 15):
+    """ selects N unique words from an image.Here, we use the url of the json, created with the barcode
+    
+    Parameters
+    ----------
+    barcode: str
+        EAN - 13 digits
+    df: pd.DataFrame
+    keys: list, default = None
```
Comment (streino) [inline make_ocr_features/utils/OCR_word_selection.py:269]: Missing param doc for `word_count_limit`

File: make_ocr_features/utils/OCR_word_selection.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    """ extracts a list with N highest tf-idf score, from a document.
+    Parameters
+    ----------
+    sub_df_idx: int
+        index of the document in the sub-dataframe (filtered dataframe on a given language).
+    feature_names:np.ndarray
+        name of the features given by vectorizer.get_feature_names_out()
+    tfidf_matrix: scipy.sparse.csr.csr_matrix
+        tfidf sparse matrix
+    N_words_to_pick: int, default = 120
```
Comment (streino) [inline make_ocr_features/utils/OCR_word_selection.py:31]: Lowercase param (to match function header)

File: make_ocr_features/utils/OCR_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            for key in keys:
+                ocr_text_dict[code][key] = json_line['ocrs'][key]['text']
+
+
+### cleaner ###
+
+def text_cleaner(text:str) -> str:
+    """ takes a text as input and cleans it"""
+
+    dont_take = ["kj", "kcal", "kj", "total", "free", "net", "ingredients", "ingredient", "et", "de", "fat", "mg", "cg", "g", "kg", "ml", "cl", "l", "kl", "per", "pour", "valeur", "or", "le", "la", "dont", "consommer", "poids", "net", "www", "com", "which", "of", "wt"]
```
Comment (streino) [inline make_ocr_features/utils/OCR_preprocessing.py:62]: `dont_take` is also defined in OCR_word_selection.py. Can it be refactored into a single definition?

File: make_ocr_features/utils/OCR_word_selection.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+    """
+    
+    if len(lang_dict) > 1:
+        sets = [set(remove_duplicates(text_cleaner(lang_dict[lang]["text"]), get_keys = True).split()) for lang in lang_dict.keys()]
+        intersection =  " ".join([word for word in set.intersection(*sets) if word not in dont_take][:n_max])
+        return intersection
+    else: 
+        return ""
+
+def get_intersect_words_ocr(sub_ocr_text_dict, N_max = 30):
```
Comment (streino) [inline make_ocr_features/utils/OCR_word_selection.py:113]: Pep8: lowercase params

**Review by streino — COMMENTED** [review]
I added a few new misc comments, and there are a few from my previous review that you didn't close/comment (esp. https://github.com/openfoodfacts/off-category-classification/pull/75#issuecomment-1163199133), but overall looks good to me.

I won't be able to review in the next 7-10 days, so if you need more feedback maybe @alexgarel can help?

**Review by raphael0202 — APPROVED** [review]
Merging this branch, as all changes are in a distinct directory. We can perform some fixes/updates later.

**Pull request merged**